storage/pebble/blocks.go

```diff
@@ -0,0 +1,180 @@
+package pebble
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/onflow/flow-evm-gateway/storage"
+	errs "github.com/onflow/flow-evm-gateway/storage/errors"
+	"github.com/onflow/flow-go/fvm/evm/types"
+)
+
+var _ storage.BlockIndexer = &Blocks{}
+
+type BlockOption func(block *Blocks) error
+
+// WithInitHeight sets the first and last height to the provided value,
+// this should be used to initialize an empty database, if the first and last
+// heights are already set an error will be returned.
+func WithInitHeight(height uint64) BlockOption {
+	return func(block *Blocks) error {
+		return block.storeInitHeight(height)
+	}
+}
+
+type Blocks struct {
+	store       *Storage
+	mux         sync.RWMutex
+	heightCache map[byte]uint64
+}
+
+func NewBlocks(store *Storage, opts ...BlockOption) (*Blocks, error) {
+	blk := &Blocks{
+		store:       store,
+		mux:         sync.RWMutex{},
+		heightCache: make(map[byte]uint64),
+	}
+
+	for _, opt := range opts {
+		if err := opt(blk); err != nil {
+			return nil, err
+		}
+	}
+
+	return blk, nil
+}
+
+func (b *Blocks) Store(block *types.Block) error {
+	b.mux.Lock()
+	defer b.mux.Unlock()
+
+	val, err := block.ToBytes()
+	if err != nil {
+		return err
+	}
+
+	id, err := block.Hash()
+	if err != nil {
+		return err
+	}
+
+	// todo batch operations
+	height := uint64Bytes(block.Height)
+	if err := b.store.set(blockHeightKey, height, val); err != nil {
+		return err
+	}
+
+	// todo check if what is more often used block by id or block by height and fix accordingly if needed
+	if err := b.store.set(blockIDToHeightKey, id.Bytes(), height); err != nil {
+		return err
+	}
+
+	return b.setLastHeight(block.Height)
+}
+
+func (b *Blocks) GetByHeight(height uint64) (*types.Block, error) {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+
+	first, err := b.FirstHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	last, err := b.LatestHeight()
+	if err != nil {
+		return nil, err
+	}
+
+	// check if the requested height is within the known range
```

Comment:
I would assume that someone running the EVM gateway, would always start the service from the height the EVM contract was deployed on a given network. Otherwise, the gateway will have indexed partial data, resulting in inconsistent state. What is the actual use of keeping track of first height? :thinking:

Reply:
This is just an extra check, it might come handy when we spork the network. It might be that the evm gateway can then only support heights above a certain evm height for execution since the previous spork might be an issue to access. I can see your point but we can still remove later if we don't need it. Better safe than sorry